### PR TITLE
query traversal improvements

### PR DESCRIPTION
**Pull request by andimarek**
- Change the query traverser to actually traverse all Nodes in the query and provide the TraverserContext to the QueryVisitorFieldEnvironment which allows access to the list of all parent nodes.
- add visitDepthFirst to QueryTraversal 
- make QueryTraversal and NodeTraverser return the accumulate
- Improves the TraverserContext by making the child contexts available for the `leave` callback.

### CODE REVIEW

File: src/main/java/graphql/analysis/QueryTraversal.java [hunk 1]
```diff
@@ -69,9 +70,8 @@ private QueryTraversal(GraphQLSchema schema,
         this.schema = assertNotNull(schema, "schema can't be null");
         this.variables = assertNotNull(variables, "variables can't be null");
         this.fragmentsByName = getOperationResult.fragmentsByName;
-        this.roots = getOperationResult.operationDefinition.getSelectionSet().getChildren();
+        this.roots = Collections.singletonList(getOperationResult.operationDefinition);
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/QueryTraversal.java]: I am a fan of static imports for readability

File: src/main/java/graphql/analysis/QueryTraversal.java [hunk 2]
```diff
@@ -163,20 +166,31 @@ private GraphQLObjectType getRootTypeFromOperation(OperationDefinition operation
...
+    private List<Node> childrenOf(Node<?> node) {
+        if (!(node instanceof FragmentSpread)) {
+            return node.getChildren();
+        }
+        FragmentSpread fragmentSpread = (FragmentSpread) node;
+        return Collections.singletonList(fragmentsByName.get(fragmentSpread.getName()));
     }
 
-    private void visitImpl(QueryVisitor visitFieldCallback, boolean preOrder) {
+    private Object visitImpl(QueryVisitor visitFieldCallback, Boolean preOrder) {
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/QueryTraversal.java]: why use null boolean as a flag.  Surely a boolean is true of false??

File: src/main/java/graphql/util/TraverserState.java [hunk 3]
```diff
@@ -56,21 +65,27 @@ private QueueTraverserState(Object sharedContextData) {
...
+            super.state.add(endList);
             super.state.add(traverserContext);
         }
     }
 
-    public enum Marker {
-        END_LIST
+    public static class EndList<U> {
+        public Map<String, List<TraverserContext<U>>> childrenContextMap;
     }
```
Comment (bbakerman) [inline src/main/java/graphql/util/TraverserState.java]: whats up with this class??   A static class with a field with public access???  Not great API

File: src/main/java/graphql/analysis/QueryTraversal.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/analysis/QueryTraversal.java]: I need three states, therefore Boolean

File: src/main/java/graphql/analysis/QueryTraversal.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/analysis/QueryTraversal.java]: changed

**Pull request merged**